docs/cudf/source/developer_guide/testing.md

```diff
@@ -0,0 +1,143 @@
+Test coverage is measured using [`coverage.py`](https://coverage.readthedocs.io/en/6.4.1/),
+specifically the [`pytest-cov`](https://github.com/pytest-dev/pytest-cov) plugin.
+Code coverage reports are uploaded to [Codecov](https://app.codecov.io/gh/rapidsai/cudf).
+Each PR also indicates whether it increases or decreases in test coverage.
+
+## Test organization
+
+How tests are organized depends on which of the following two groups they fall into:
+
+1. Free functions operating on the above classes like `cudf.merge`.
+2. Methods of classes like `DataFrame` or `Series`.
+
+Tests of free functions should be grouped into files based on the API sections in our documentation.
+This places tests of similar functionality into the same testing module.
+Tests of class methods should be organized in the same way, except that this organization should be within a subdirectory corresponding to the class.
+For instance, tests of `DataFrame` indexing should be placed into `dataframe/test_indexing.py`.
+In cases where tests may be shared by multiple classes sharing a common parent (e.g. `DataFrame` and `Series` both require `IndexedFrame` tests),
+the tests may be placed in a directory corresponding to the parent class.
+
+## Test contents
+
+### Writing tests
+
+In general, functionality must be tested both for standard use cases and exceptional cases.
+Standard use cases may be covered using parametrization (using `pytest.mark.parametrize`).
+Tests of standard use cases should typically include some coverage of:
+- Different dtypes, including nested dtypes (especially strings)
+- Mixed objects, e.g. binary operations between `DataFrame` and `Series`
+- Operations on scalars
+- Verifying all combinations of parameters for complex APIs like `cudf.merge`.
+
+Here are some of the most common exceptional cases to test:
+1. `Series`/`DataFrame`/`Index` with zero rows
+2. `DataFrame` with zero columns
+3. All null data
+4. For string or list APIs, empty strings/lists
+5. For list APIs, lists containing all null elements or empty strings
+6. For numeric data:
+  1. All 0s.
+  2. All 1s.
+  3. Containing/all inf
+  4. Containing/all nan
+  5. `INT${PRECISION}_MAX` for a given precision (e.g. `2**32` for `int32`).
+7. For strings, non-`en_US` locale.
+
+Most specific APIs will also include a range of other cases.
+
+In general, it is preferable to write separate tests for different exception cases.
+Excessive parametrization and branching increases complexity and obfuscates the purpose of a test.
+Typically, exception cases require specific assertions or other special logic, so they are best kept separate.
+The main nexception to this rule is tests based on comparison to pandas.
+Such tests may test exceptional cases alongside more typical cases since the logic is generally identical.
+
+### Parametrization: custom fixtures and `pytest.mark.parametrize`
+
+When it comes to parametrizing tests, the two main options are fixtures and `mark.parametrize`.
```

Comment:
```suggestion
When it comes to parametrizing tests, the two main options are pytest fixtures and `pytest.mark.parametrize`.
```

Reply:
Link to:

- fixtures: https://docs.pytest.org/en/stable/explanation/fixtures.html
- parametrize: https://docs.pytest.org/en/stable/how-to/parametrize.html#pytest-mark-parametrize

?

Reply:
Rather than duplicating `pytest`, I've worded this as "...parametrizing tests using `pytest`, the two main options are ...". And yes, I'll add links.